From d1ee7c6db92276eeb8bf47d22b1e0c6804b1e7ac Mon Sep 17 00:00:00 2001
From: Alexandre Fayolle <alexandre.fayolle@camptocamp.com>
Date: Tue, 29 Oct 2019 14:46:24 +0100
Subject: [PATCH] [FIX] 10.0 sale_triple_discount

there are cases where the cache trick done in triple_discount_preprocess
is lost when sale.order.line _compute_amount is called and processes the
1st line of the sale order => all remaining lines are impacted, and don't
have the discounts 2 and 3 taken into account.

Not updating the sale.order _amount_all method because it is unlikely that
fields on the lines depend on the total amount of the sale.
---
 sale_triple_discount/README.rst                |  1 +
 sale_triple_discount/__manifest__.py           |  2 +-
 sale_triple_discount/models/sale_order_line.py | 13 ++++++++++---
 3 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/sale_triple_discount/README.rst b/sale_triple_discount/README.rst
index a9e14cf9..16c9efd4 100644
--- a/sale_triple_discount/README.rst
+++ b/sale_triple_discount/README.rst
@@ -80,6 +80,7 @@ Contributors
 * David Vidal <david.vidal@tecnativa.com>
 * Simone Rubino <simone.rubino@agilebg.com>
 * Jacques-Etienne Baudoux (BCIM sprl) <je@bcim.be>
+* Alexandre Fayolle <alexandre.fayolle@camptocamp.com>
 
 Maintainer
 ----------
diff --git a/sale_triple_discount/__manifest__.py b/sale_triple_discount/__manifest__.py
index 96ab9ec9..791b1aeb 100644
--- a/sale_triple_discount/__manifest__.py
+++ b/sale_triple_discount/__manifest__.py
@@ -7,7 +7,7 @@
 
 {
     'name': 'Sale Triple Discount',
-    'version': '10.0.1.0.1',
+    'version': '10.0.1.1.0',
     'category': 'Sales Management',
     'author': 'ADHOC SA, '
               'Agile Business Group, '
diff --git a/sale_triple_discount/models/sale_order_line.py b/sale_triple_discount/models/sale_order_line.py
index a8c13aa5..98ae7087 100644
--- a/sale_triple_discount/models/sale_order_line.py
+++ b/sale_triple_discount/models/sale_order_line.py
@@ -49,9 +49,16 @@ class SaleOrderLine(models.Model):
 
     @api.depends('discount2', 'discount3', 'discounting_type')
     def _compute_amount(self):
-        prev_values = self.triple_discount_preprocess()
-        super(SaleOrderLine, self)._compute_amount()
-        self.triple_discount_postprocess(prev_values)
+        # we iterate and work line by line, because _compute_amount calls
+        # update, which can invalidate the cache, depending on the presence of
+        # other fields depending on the ones computed by that method. In that
+        # case the work done in preprocess would be lost as soon as the 1st
+        # line is updated, causing the following lines to ignore the discount2
+        # and discount3 fields.
+        for line in self:
+            prev_values = line.triple_discount_preprocess()
+            super(SaleOrderLine, line)._compute_amount()
+            line.triple_discount_postprocess(prev_values)
 
     discount2 = fields.Float(
         'Disc. 2 (%)',
-- 
2.17.1

